### PR TITLE
[FL-2529][FL-1628] New LF-RFID subsystem

### PR DESCRIPTION
**Pull request by DrZlo13**
# What's new

New protocols:
- Indala26 (covers almost all Indala PSK cards)
- AWID
- FDX-A
- FDX-B
- 96-bit HID protocol (covers up to HID 40-bit)
- 192-bit HID protocol (covers HID extended)
- RAW RFID data recording

# Verification 

- Use the LF-RFID application as usual, try to read, emulate, write new protocols.

# Checklist (For Reviewer)

- [ ] PR has description of feature/bug or link to Confluence/Jira task
- [ ] Description contains actions to verify feature/bugfix
- [ ] I've built this code, uploaded it to the device and verified feature/bugfix

### CODE REVIEW

File: applications/archive/helpers/archive_apps.c [hunk 1]
```diff
@@ -29,23 +29,13 @@ bool archive_app_is_available(void* context, const char* path) {
...
-        if(file_exists) {
-            storage_file_close(file);
-            file_exists =
-                storage_file_open(file, ANY_PATH("u2f/cnt.u2f"), FSAM_READ, FSOM_OPEN_EXISTING);
-            if(file_exists) {
-                storage_file_close(file);
-            }
+        Storage* storage = furi_record_open(RECORD_STORAGE);
+
+        if(storage_file_exists(storage, "/any/u2f/key.u2f")) {
```
Comment (hedger) [inline applications/archive/helpers/archive_apps.c]: ANY_PATH("u2f/key.u2f")

File: applications/lfrfid/lfrfid_app.cpp [hunk 2]
```diff
@@ -37,15 +37,35 @@ LfRfidApp::LfRfidApp()
     , storage{"storage"}
```
Comment (hedger) [inline applications/lfrfid/lfrfid_app.cpp]: RECORD_STORAGE

File: applications/lfrfid/lfrfid_app.cpp [hunk 3]
```diff
@@ -37,15 +37,35 @@ LfRfidApp::LfRfidApp()
...
     , dialogs{"dialogs"}
```
Comment (hedger) [inline applications/lfrfid/lfrfid_app.cpp]: RECORD_DIALOGS

File: applications/lfrfid/lfrfid_cli.c [hunk 4]
```diff
@@ -0,0 +1,643 @@
...
+
+#define RAW_HEADER "RFID RAW"
+
+#define FURI_SW_MEMBARRIER() asm volatile("" : : : "memory")
+
+static void lfrfid_cli(Cli* cli, string_t args, void* context);
+
+// app cli function
+void lfrfid_on_system_start() {
+    Cli* cli = furi_record_open("cli");
```
Comment (hedger) [inline applications/lfrfid/lfrfid_cli.c]: RECORD_CLI

File: applications/lfrfid/lfrfid_cli.c [hunk 5]
```diff
@@ -0,0 +1,643 @@
...
+    cli_add_command(cli, "rfid", CliCommandFlagDefault, lfrfid_cli, NULL);
+    furi_record_close("cli");
```
Comment (hedger) [inline applications/lfrfid/lfrfid_cli.c]: RECORD_CLI

File: applications/lfrfid/lfrfid_cli.c [hunk 6]
```diff
@@ -0,0 +1,643 @@
...
+
+        protocol_dict_free(dict);
+        free(buffer);
+
+    } while(false);
+
+    stream_free(stream);
+    string_clear(filepath);
+    string_clear(info_string);
+    furi_record_close("storage");
```
Comment (hedger) [inline applications/lfrfid/lfrfid_cli.c]: RECORD_STORAGE

File: applications/lfrfid/lfrfid_cli.c [hunk 7]
```diff
@@ -0,0 +1,643 @@
...
+    lfrfid_worker_free(worker);
+    protocol_dict_free(dict);
+}
+
+static void lfrfid_cli_raw_analyze(Cli* cli, string_t args) {
+    UNUSED(cli);
+    string_t filepath, info_string;
+    string_init(filepath);
+    string_init(info_string);
+    Storage* storage = furi_record_open("storage");
```
Comment (hedger) [inline applications/lfrfid/lfrfid_cli.c]: RECORD_STORAGE

File: applications/lfrfid/lfrfid_cli.c [hunk 8]
```diff
@@ -0,0 +1,643 @@
...
+    }
+    printf("Emulation stopped\r\n");
+
+    lfrfid_worker_stop(worker);
+    lfrfid_worker_stop_thread(worker);
+    lfrfid_worker_free(worker);
+    protocol_dict_free(dict);
+}
+
+static void lfrfid_cli_raw_analyze(Cli* cli, string_t args) {
```
Comment (hedger) [inline applications/lfrfid/lfrfid_cli.c:264]: 150+ lines function

File: applications/lfrfid/lfrfid_cli.c [hunk 9]
```diff
@@ -0,0 +1,643 @@
...
+    FuriEventFlag* event = context;
+    furi_event_flag_set(event, 1 << result);
+}
+
+static void lfrfid_cli_raw_emulate(Cli* cli, string_t args) {
+    UNUSED(cli);
+
+    string_t filepath;
+    string_init(filepath);
+    Storage* storage = furi_record_open("storage");
```
Comment (hedger) [inline applications/lfrfid/lfrfid_cli.c]: RECORD_STORAGE

File: applications/lfrfid/lfrfid_cli.c [hunk 10]
```diff
@@ -0,0 +1,643 @@
...
+
+        lfrfid_worker_stop_thread(worker);
+        lfrfid_worker_free(worker);
+        protocol_dict_free(dict);
+
+        furi_event_flag_free(event);
+
+    } while(false);
+
+    furi_record_close("storage");
```
Comment (hedger) [inline applications/lfrfid/lfrfid_cli.c]: RECORD_STORAGE

File: applications/lfrfid/scene/lfrfid_app_scene_read_success.cpp [hunk 11]
```diff
@@ -4,10 +4,37 @@
...
+    for(uint8_t i = 0; i < size; i++) {
+        if(i != 0) {
+            string_cat_printf(string_info, " ");
+        }
+
+        if(i >= 9) {
+            string_cat_printf(string_info, "...");
+            break;
+        } else {
+            string_cat_printf(string_info, "%02X", data[i]);
```
Comment (hedger) [inline applications/lfrfid/scene/lfrfid_app_scene_read_success.cpp:28]: move to a helper function?

File: lib/lfrfid/lfrfid_raw_worker.c [hunk 12]
```diff
@@ -0,0 +1,602 @@
...
+    furi_thread_set_callback(worker->thread, lfrfid_raw_read_worker_thread);
+
+    furi_thread_start(worker->thread);
+    return result;
+}
+
+bool lfrfid_raw_worker_start_emulate(LFRFIDRawWorker* worker, const char* file_path) {
+    furi_check(furi_thread_get_state(worker->thread) == FuriThreadStateStopped);
+
+    bool result = true;
```
Comment (hedger) [inline lib/lfrfid/lfrfid_raw_worker.c]: unused

File: lib/lfrfid/lfrfid_raw_worker.c [hunk 13]
```diff
@@ -0,0 +1,602 @@
...
+    string_set(worker->file_path, file_path);
+
+    furi_thread_set_callback(worker->thread, lfrfid_raw_emulate_worker_thread);
+
+    furi_thread_start(worker->thread);
+    return result;
+}
+
+bool lfrfid_raw_worker_stop(LFRFIDRawWorker* worker) {
+    bool result = true;
```
Comment (hedger) [inline lib/lfrfid/lfrfid_raw_worker.c]: unused

File: lib/lfrfid/lfrfid_raw_worker.c [hunk 14]
```diff
@@ -0,0 +1,602 @@
...
+            size_t s = xStreamBufferSendFromISR(
+                ctx->stream, &buffer_p, sizeof(RfidReadBuffer*), &xHigherPriorityTaskWoken);
+
+            if(s != sizeof(RfidReadBuffer*)) {
+                ctx->overrun_count++;
+            }
+        }
+
+        if(!ctx->buffers[id].occupied) {
+            memcpy(
```
Comment (hedger) [inline lib/lfrfid/lfrfid_raw_worker.c]: move block to separate function?

File: lib/lfrfid/lfrfid_raw_worker.c [hunk 15]
```diff
@@ -0,0 +1,602 @@
...
+        }
+    }
+
+    portYIELD_FROM_ISR(xHigherPriorityTaskWoken);
+}
+
+static int32_t lfrfid_raw_read_worker_thread(void* thread_context) {
+    LFRFIDRawWorker* worker = (LFRFIDRawWorker*)thread_context;
+
+    Storage* storage = furi_record_open("storage");
```
Comment (hedger) [inline lib/lfrfid/lfrfid_raw_worker.c]: RECORD_STORAGE

File: lib/lfrfid/lfrfid_raw_worker.c [hunk 16]
```diff
@@ -0,0 +1,602 @@
...
+
+            if(FURI_BIT(flags, LFRFIDRawWorkerEventStop)) {
+                break;
+            };
+        }
+    }
+
+    vStreamBufferDelete(data->ctx.stream);
+    stream_free(stream);
+    furi_record_close("storage");
```
Comment (hedger) [inline lib/lfrfid/lfrfid_raw_worker.c]: RECORD_STORAGE

File: lib/lfrfid/lfrfid_raw_worker.c [hunk 17]
```diff
@@ -0,0 +1,602 @@
...
+
+    return 0;
+}
+
+static void rfid_emulate_dma_isr(bool half, void* context) {
+    RfidEmulateCtx* ctx = context;
+
+    if(half) {
+        uint32_t flag = HalfTransfer;
+        size_t len = xStreamBufferSendFromISR(ctx->stream, &flag, sizeof(uint32_t), pdFALSE);
```
Comment (hedger) [inline lib/lfrfid/lfrfid_raw_worker.c]: copypasted

Comment (hedger) [reply lib/lfrfid/lfrfid_raw_worker.c]: move outside?

File: lib/lfrfid/lfrfid_raw_worker.c [hunk 18]
```diff
@@ -0,0 +1,602 @@
...
+}
+
+static int32_t lfrfid_raw_emulate_worker_thread(void* thread_context) {
+    LFRFIDRawWorker* worker = thread_context;
+
+    bool file_valid = true;
+
+    LFRFIDRawWorkerEmulateData* data = malloc(sizeof(LFRFIDRawWorkerEmulateData));
+
+    Storage* storage = furi_record_open("storage");
```
Comment (hedger) [inline lib/lfrfid/lfrfid_raw_worker.c]: RECORD_STORAGE

File: lib/lfrfid/lfrfid_raw_worker.c [hunk 19]
```diff
@@ -0,0 +1,602 @@
...
+                }
+
+                size = stream_write(stream, buffer_p->data, buffer_p->size);
+                if(size != buffer_p->size) {
+                    file_valid = false;
+                }
+
+                buffer_p->size = 0;
+                __DMB();
+                buffer_p->occupied = false;
```
Comment (hedger) [inline lib/lfrfid/lfrfid_raw_worker.c]: volatile?

File: lib/lfrfid/lfrfid_raw_worker.c [hunk 20]
```diff
@@ -0,0 +1,602 @@
...
+
+        size_t size = stream_read(data->stream, (uint8_t*)header_text, strlen(RAW_HEADER));
+        if(size != strlen(RAW_HEADER) || strcmp(header_text, RAW_HEADER) != 0) {
+            file_valid = false;
+        }
+
+        free(header_text);
+
+        uint32_t max_buffer_size;
+        size = stream_read(data->stream, (uint8_t*)&max_buffer_size, sizeof(uint32_t));
```
Comment (hedger) [inline lib/lfrfid/lfrfid_raw_worker.c]: same code in CLI loader, move to 1 place?

File: lib/lfrfid/lfrfid_raw_worker.c [hunk 21]
```diff
@@ -0,0 +1,602 @@
...
+        }
+    }
+
+    if(data->ctx.overrun_count) {
+        FURI_LOG_E(TAG_EMULATE, "overruns: %lu", data->ctx.overrun_count);
+    }
+
+    stream_free(data->stream);
+    vStreamBufferDelete(data->ctx.stream);
+    furi_record_close("storage");
```
Comment (hedger) [inline lib/lfrfid/lfrfid_raw_worker.c]: RECORD_STORAGE

File: lib/lfrfid/lfrfid_worker_modes.c [hunk 22]
```diff
@@ -0,0 +1,514 @@
...
+    uint32_t timeout,
+    ProtocolId* result_protocol) {
+    LFRFIDWorkerReadState state;
+    furi_hal_rfid_pins_read();
+
+    if(feature & LFRFIDFeatureASK) {
+        furi_hal_rfid_tim_read(125000, 0.5);
+        FURI_LOG_D(TAG, "Start ASK");
+    } else {
+        furi_hal_rfid_tim_read(62500, 0.25);
```
Comment (hedger) [inline lib/lfrfid/lfrfid_worker_modes.c:118]: magic

File: lib/lfrfid/lfrfid_worker_modes.c [hunk 23]
```diff
@@ -0,0 +1,514 @@
...
+    HalfTransfer,
+    TransferComplete,
+} LFRFIDWorkerEmulateDMAEvent;
+
+static void lfrfid_worker_emulate_dma_isr(bool half, void* context) {
+    StreamBufferHandle_t stream = context;
+
+    if(half) {
+        uint32_t flag = HalfTransfer;
+        xStreamBufferSendFromISR(stream, &flag, sizeof(uint32_t), pdFALSE);
```
Comment (hedger) [inline lib/lfrfid/lfrfid_worker_modes.c]: move out of if()

File: lib/lfrfid/lfrfid_worker_modes.c [hunk 24]
```diff
@@ -0,0 +1,514 @@
...
+/**************************************************************************************************/
+
+void lfrfid_worker_mode_read_raw_process(LFRFIDWorker* worker) {
+    LFRFIDRawWorker* raw_worker = lfrfid_raw_worker_alloc();
+
+    lfrfid_raw_worker_read_set_callback(raw_worker, worker->read_raw_cb, worker->cb_ctx);
+
+    switch(worker->read_type) {
+    case LFRFIDWorkerReadTypePSKOnly:
+        lfrfid_raw_worker_start_read(raw_worker, worker->raw_filename, 62500, 0.25);
```
Comment (hedger) [inline lib/lfrfid/lfrfid_worker_modes.c]: magic #2

File: lib/lfrfid/protocols/lfrfid_protocols.h [hunk 25]
```diff
@@ -0,0 +1,32 @@
...
+
+extern const ProtocolBase* lfrfid_protocols[];
+
+typedef enum {
+    LFRFIDWriteTypeT5577,
+} LFRFIDWriteType;
+
+typedef struct {
+    LFRFIDWriteType write_type;
+    union {
```
Comment (hedger) [inline lib/lfrfid/protocols/lfrfid_protocols.h:32]: union?

File: lib/lfrfid/protocols/protocol_awid.c [hunk 26]
```diff
@@ -0,0 +1,239 @@
...
+        request->t5577.block[3] = bit_lib_get_bits_32(protocol->encoded_data, 64, 32);
+        request->t5577.blocks_to_write = 4;
+        result = true;
+    }
+    return result;
+};
+
+const ProtocolBase protocol_awid = {
+    .name = "AWID",
+    .manufacturer = "AWIG",
```
Comment (hedger) [inline lib/lfrfid/protocols/protocol_awid.c:219]: awid?

File: lib/lfrfid/protocols/protocol_h10301.c [hunk 27]
```diff
@@ -0,0 +1,386 @@
...
+    protocol_h10301_write_raw_bit(0, 1, card_data);
+    protocol_h10301_write_raw_bit(0, 2, card_data);
+    protocol_h10301_write_raw_bit(1, 3, card_data);
+    protocol_h10301_write_raw_bit(1, 4, card_data);
+    protocol_h10301_write_raw_bit(1, 5, card_data);
+    protocol_h10301_write_raw_bit(0, 6, card_data);
+    protocol_h10301_write_raw_bit(1, 7, card_data);
+
+    // company / OEM code 1
+    protocol_h10301_write_bit(0, 8, card_data);
```
Comment (hedger) [inline lib/lfrfid/protocols/protocol_h10301.c:257]: maybe use a cycle?

File: lib/lfrfid/protocols/protocol_io_prox_xsf.c [hunk 28]
```diff
@@ -0,0 +1,297 @@
...
+    // Validate the packet preamble is there...
+    if(encoded_data[0] != 0b00000000) {
+        return false;
+    }
+    if((encoded_data[1] >> 6) != 0b01) {
+        return false;
+    }
+
+    // ... check for known ones...
+    if(bit_lib_bit_is_not_set(encoded_data[2], 6)) {
```
Comment (hedger) [inline lib/lfrfid/protocols/protocol_io_prox_xsf.c:98]: use cycle?

File: lib/toolbox/protocols/protocol_dict.c [hunk 29]
```diff
@@ -0,0 +1,226 @@
...
+    uint32_t duration) {
+    bool done = false;
+    ProtocolId ready_protocol_id = PROTOCOL_NO;
+
+    for(size_t i = 0; i < dict->count; i++) {
+        uint32_t features = dict->base[i]->features;
+        if(features & feature) {
+            ProtocolDecoderFeed fn = dict->base[i]->decoder.feed;
+
+            if(fn) {
```
Comment (hedger) [inline lib/toolbox/protocols/protocol_dict.c:133]: move to single condition

File: lib/toolbox/pulse_joiner.c [hunk 30]
```diff
@@ -0,0 +1,117 @@
...
+void pulse_joiner_free(PulseJoiner* pulse_joiner) {
+    free(pulse_joiner);
+}
+
+bool pulse_joiner_push_pulse(PulseJoiner* pulse_joiner, bool polarity, size_t period, size_t pulse) {
+    bool result = false;
+    furi_check((pulse_joiner->pulse_index + 1) < PULSE_MAX_COUNT);
+
+    if(polarity == false && pulse_joiner->pulse_index == 0) {
+        // first negative pulse is ommited
```
Comment (hedger) [inline lib/toolbox/pulse_joiner.c]: omitted

File: lib/toolbox/varint.h [hunk 31]
```diff
@@ -0,0 +1,27 @@
+#pragma once
+#include <stdint.h>
+#include <stdlib.h>
+
```
Comment (hedger) [inline lib/toolbox/varint.h:4]: c++ header wrapper?

File: lib/lfrfid/tools/fsk_demod.h [hunk 32]
```diff
@@ -0,0 +1,12 @@
+#pragma once
+#include <stdint.h>
+#include <stdbool.h>
+
+typedef struct FSKDemod FSKDemod;
+
```
Comment (hedger) [inline lib/lfrfid/tools/fsk_demod.h:10]: c++ header wrapper?

File: lib/toolbox/profiler.h [hunk 33]
```diff
@@ -0,0 +1,15 @@
+#pragma once
+
```
Comment (hedger) [inline lib/toolbox/profiler.h:2]: c++ header wrapper?

File: lib/toolbox/pulse_joiner.h [hunk 34]
```diff
@@ -0,0 +1,38 @@
+#include <stdint.h>
+#include <stdbool.h>
+#include <stddef.h>
+
```
Comment (hedger) [inline lib/toolbox/pulse_joiner.h:4]: c++ header wrapper?

**Review by hedger — CHANGES_REQUESTED** [review]
Remove .vscode.zip

File: lib/lfrfid/lfrfid_worker_modes.c (hunk 22 above)
Comment (DrZlo13) [reply lib/lfrfid/lfrfid_worker_modes.c:118]: Literally frequency and duty cycle, as furi_hal_rfid_tim_read arguments says. I don't see the point in naming it.

File: lib/lfrfid/lfrfid_worker_modes.c (hunk 24 above)
Comment (DrZlo13) [reply lib/lfrfid/lfrfid_worker_modes.c]: Literally frequency and duty cycle, as furi_hal_rfid_tim_read arguments says. I don't see the point in naming it.

File: lib/lfrfid/protocols/lfrfid_protocols.h (hunk 25 above)
Comment (DrZlo13) [reply lib/lfrfid/protocols/lfrfid_protocols.h:32]: To make it easier to extend the code later. We want not only T5577.

File: lib/lfrfid/protocols/protocol_awid.c (hunk 26 above)
Comment (DrZlo13) [reply lib/lfrfid/protocols/protocol_awid.c:219]: Applied Wireless Identifications Data by Applied Wireless Identifications Group, iirc

File: lib/toolbox/protocols/protocol_dict.c (hunk 29 above)
Comment (DrZlo13) [reply lib/toolbox/protocols/protocol_dict.c:133]: will be less readable

File: lib/lfrfid/protocols/protocol_io_prox_xsf.c (hunk 28 above)
Comment (DrZlo13) [reply lib/lfrfid/protocols/protocol_io_prox_xsf.c:98]: will look bad

File: lib/lfrfid/protocols/protocol_h10301.c (hunk 27 above)
Comment (DrZlo13) [reply lib/lfrfid/protocols/protocol_h10301.c:257]: h10301 needs rework, it will be done later

File: lib/lfrfid/lfrfid_raw_worker.c (hunk 19 above)
Comment (DrZlo13) [reply lib/lfrfid/lfrfid_raw_worker.c]: volatile will load the cpu more than we can afford, this place is extremely sensitive to that

File: applications/lfrfid/lfrfid_cli.c (hunk 8 above)
Comment (DrZlo13) [reply applications/lfrfid/lfrfid_cli.c:264]: Was simplified

File: lib/lfrfid/lfrfid_raw_worker.c (hunk 14 above)
Comment (DrZlo13) [reply lib/lfrfid/lfrfid_raw_worker.c]: Moved

File: lib/lfrfid/lfrfid_raw_worker.c (hunk 20 above)
Comment (DrZlo13) [reply lib/lfrfid/lfrfid_raw_worker.c]: Moved

**Comment by exussum12** [comment]
lib/littlefs/lfs.c:1070: undefined reference to `furi_log_print'

I'm getting that when compiling

File: lib/lfrfid/lfrfid_raw_file.c [hunk 35]
```diff
@@ -0,0 +1,149 @@
...
+    uint32_t max_buffer_size) {
+    LFRFIDRawFileHeader header = {
+        .magic = LFRFID_RAW_FILE_MAGIC,
+        .version = LFRFID_RAW_FILE_VERSION,
+        .frequency = frequency,
+        .duty_cycle = duty_cycle,
+        .max_buffer_size = max_buffer_size};
+
+    size_t size = stream_write(file->stream, (uint8_t*)&header, sizeof(LFRFIDRawFileHeader));
+    if(size == sizeof(LFRFIDRawFileHeader)) {
```
Comment (hedger) [inline lib/lfrfid/lfrfid_raw_file.c]: return (size == sizeof(LFRFIDRawFileHeader));

File: lib/lfrfid/lfrfid_raw_file.c [hunk 36]
```diff
@@ -0,0 +1,149 @@
...
+
+bool lfrfid_raw_file_write_buffer(LFRFIDRawFile* file, uint8_t* buffer_data, size_t buffer_size) {
+    size_t size;
+    size = stream_write(file->stream, (uint8_t*)&buffer_size, sizeof(size_t));
+    if(size != sizeof(size_t)) return false;
+
+    size = stream_write(file->stream, buffer_data, buffer_size);
+    if(size != buffer_size) return false;
+
+    return true;
```
Comment (hedger) [inline lib/lfrfid/lfrfid_raw_file.c:73]: return (stream_write() == sizeof()) && (stream_write() == buffer_size)
would be easier to read. Personal opinion, completely optional

File: lib/lfrfid/protocols/protocol_fdx_a.c [hunk 37]
```diff
@@ -0,0 +1,243 @@
...
+    uint8_t decoded_data[FDXA_DECODED_DATA_SIZE];
+    if(!protocol_fdx_a_decode(data, decoded_data)) return false;
+
+    uint8_t parity_sum = 0;
+    for(size_t i = 0; i < FDXA_DECODED_DATA_SIZE; i++) {
+        parity_sum += bit_lib_test_parity_32(decoded_data[i], BitLibParityOdd);
+        decoded_data[i] &= 0x7F;
+    }
+
+    if(parity_sum == 0) {
```
Comment (hedger) [inline lib/lfrfid/protocols/protocol_fdx_a.c]: return (parity_sum == 0)

File: lib/toolbox/buffer_stream.c [hunk 38]
```diff
@@ -0,0 +1,145 @@
...
+    size_t size,
+    BaseType_t* const task_woken) {
+    Buffer* buffer = &buffer_stream->buffers[buffer_stream->index];
+    bool result = true;
+
+    // write to buffer
+    if(!buffer_write(buffer, data, size)) {
+        // if buffer is full - send it
+        buffer->occupied = true;
+        // we always has space for buffer in stream
```
Comment (hedger) [inline lib/toolbox/buffer_stream.c]: have

**Review by hedger — CHANGES_REQUESTED** [review]

**Comment by hedger** [comment]
@exussum12 either use pre-built verison from our CI in comments above, or run fbt with -c flag

File: lib/lfrfid/lfrfid_dict_file.c [hunk 39]
```diff
@@ -0,0 +1,181 @@
...
+               file, "Key type", protocol_dict_get_name(dict, protocol)))
+            break;
+
+        // TODO: write comment about protocol sizes into file
+
+        protocol_dict_get_data(dict, protocol, data, data_size);
+
+        if(!flipper_format_write_hex(file, "Data", data, data_size)) break;
+        result = true;
+    } while(0);
```
Comment (exussum12) [inline lib/lfrfid/lfrfid_dict_file.c]: Minor but while(0) used here, while (false) used on 173, should likely be the same

File: SConstruct [hunk 40]
```diff
@@ -168,6 +168,8 @@ copro_dist = distenv.CoproBuilder(
 distenv.Alias("copro_dist", copro_dist)
 
 firmware_flash = distenv.AddOpenOCDFlashTarget(firmware_env)
+if distenv["FORCE"]:
```
Comment (xMasterX) [inline SConstruct]: Looks like some old code left that has been removed/changed in dev

(Same thing for 2-3 files below)

See commit:
[SConstruct](https://github.com/flipperdevices/flipperzero-firmware/commit/a1637e9216e98396938debd7ccb4e15cbcb41eb1#diff-d78c5c377c03dd251edc72cdb0ae7d9986fa08361eb392187218f8c670c0bc4e)

Comment (DrZlo13) [reply SConstruct]: Good catch.

File: lib/lfrfid/lfrfid_raw_worker.c [hunk 41]
```diff
@@ -0,0 +1,357 @@
...
+            if(buffer_stream_get_overrun_count(data->stream) > 0 &&
+               worker->read_callback != NULL) {
+                // message overrun to worker
+                worker->read_callback(LFRFIDWorkerReadRawOverrun, worker->context);
+            }
+
+            uint32_t flags = furi_event_flag_get(worker->events);
+            if(FURI_BIT(flags, LFRFIDRawWorkerEventStop)) {
+                break;
+            };
```
Comment (wosk) [inline lib/lfrfid/lfrfid_raw_worker.c]:
```suggestion
            }
```

File: lib/lfrfid/lfrfid_raw_worker.c [hunk 42]
```diff
@@ -0,0 +1,357 @@
...
+
+    if(!file_valid) {
+        const uint32_t available_flags = (1 << LFRFIDRawWorkerEventStop);
+        while(true) {
+            uint32_t flags = furi_event_flag_wait(
+                worker->events, available_flags, FuriFlagWaitAny, FuriWaitForever);
+
+            if(FURI_BIT(flags, LFRFIDRawWorkerEventStop)) {
+                break;
+            };
```
Comment (wosk) [inline lib/lfrfid/lfrfid_raw_worker.c]:
```suggestion
            }
```

**Review by hedger — DISMISSED** [review]

**Review by hedger — DISMISSED** [review]

**Comment by kubrickfr** [comment]
I tried this branch.

It seems like neither of the fobs I own support the protocols implemented here. 

But, before this PR, the device LED was flashing yellow when detecting a device, now there is no visual clue at all that at least "something" is working.

**Comment by kubrickfr** [comment]
> But, before this PR, the device LED was flashing yellow when detecting a device, now there is no visual clue at all that at least "something" is working.

d8b8f74 fixed it. Thank you @DrZlo13

**Comment by Tozapid** [comment]
Unfortunately not working on my pets. They have  ISO Transponder Microchip 2.12*12mm, think this one - https://www.animal-microchip.com/sale-9568205-professional-iso-transponder-microchip-2-12-12mm-for-animals-identification.html


R/W STANDARD: | ISO 11784/5 FDX-B 134.2khz
-- | --

But with your branch there is no reaction at all. On version 0.64.3, I can see the LED indication that the chip is responding


Upd. My mistake, I had to choose Extra Actions in the menu. So, now I see that LED is blinking green(or yellow, not sure), but still not reading the microchip 

Upd2. Found the spot:) Thank you! It is works
Screenshot-20220823-155700

**Review by skotopes — APPROVED** [review]

**Review by hedger — APPROVED** [review]

**Review by skotopes — APPROVED** [review]

**Pull request merged**